### PR TITLE
infoschema: fix not null flag does not work

### PR DESCRIPTION
**Pull request by zhexuany**
This PR fixes a inconsistent behavior compared with mysql. 

For query `SELECT DISTINCT C.COLLATION_NAME FROM information_schema.TABLES AS T, information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` AS C WHERE C.collation_name = T.table_collation and T.table_schema = database();`)`:

mysql returns "utf8_bin", but tidb returns empty set.

### CODE REVIEW

File: perfschema/init.go [hunk 1]
```diff
@@ -242,7 +242,7 @@ func (ps *PerfSchema) buildModel(tbName string, colNames []string, cols []column
 	ps.tables[tbName] = &model.TableInfo{
 		Name:    model.NewCIStr(tbName),
 		Charset: "utf8",
-		Collate: "utf8",
+		Collate: "utf8_general_ci",
```
Comment (shenli) [inline perfschema/init.go]: Can we use the const [here](https://github.com/pingcap/tidb/blob/master/mysql/charset.go#L559)?

File: infoschema/tables.go [hunk 2]
```diff
@@ -637,6 +636,10 @@ func dataForTables(ctx sessionctx.Context, schemas []*model.DBInfo) [][]types.Da
 	createTimeTp := tablesCols[15].tp
 	for _, schema := range schemas {
 		for _, table := range schema.Tables {
+			collation := table.Collate
+			if collation == "" {
+				collation = charset.CollationUTF8
```
Comment (shenli) [inline infoschema/tables.go]: Check the value of charset, the decide whether use CollationUTF8MB4 or CollationUTF8.

File: perfschema/init.go (hunk 1 above)
Comment (zhexuany) [reply perfschema/init.go]: Yeah, we can do this. Let me update this change.

**Comment by zhexuany** [comment]
/run-common-test=pr/470

**Comment by zhexuany** [comment]
/run-all-test=pr/470

**Comment by zhexuany** [comment]
/run-all-tests=pr/470

**Comment by zhexuany** [comment]
@zz-jason @coocood PTAL

**Comment by zhexuany** [comment]
test failure due to failure on starting tidb server. 
Rerun test /run-integration-common-test tidb-test=pr/470

File: infoschema/tables.go [hunk 3]
```diff
@@ -637,6 +638,10 @@ func dataForTables(ctx sessionctx.Context, schemas []*model.DBInfo) [][]types.Da
 	createTimeTp := tablesCols[15].tp
 	for _, schema := range schemas {
 		for _, table := range schema.Tables {
+			collation := table.Collate
+			if collation == "" {
```
Comment (zimulala) [inline infoschema/tables.go]: Which schema table will be in this situation？

Comment (zhexuany) [reply infoschema/tables.go]: Could be more specific? I don't think I fully understand your question.

Comment (zimulala) [reply infoschema/tables.go]: What I mean is which table's charset is not empty but collate is empty. Could you give an example?

File: util/charset/charset.go [hunk 4]
```diff
@@ -117,6 +104,9 @@ func ValidCharsetAndCollation(cs string, co string) bool {
 
 // GetDefaultCollation returns the default collation for charset.
 func GetDefaultCollation(charset string) (string, error) {
+	if charset == "" {
```
Comment (coocood) [inline util/charset/charset.go]: I think we better not add this logic in this function.

File: infoschema/tables.go [hunk 5]
```diff
@@ -637,6 +640,10 @@ func dataForTables(ctx sessionctx.Context, schemas []*model.DBInfo) [][]types.Da
...
+				collation, _ = charset.GetDefaultCollation(table.Charset)
```
Comment (coocood) [inline infoschema/tables.go]: We can check if `table.Charset` is empty string here.

File: infoschema/tables.go [hunk 6]
```diff
@@ -137,7 +139,7 @@ var tablesCols = []columnInfo{
 	{"CREATE_TIME", mysql.TypeDatetime, 19, 0, nil, nil},
 	{"UPDATE_TIME", mysql.TypeDatetime, 19, 0, nil, nil},
 	{"CHECK_TIME", mysql.TypeDatetime, 19, 0, nil, nil},
-	{"TABLE_COLLATION", mysql.TypeVarchar, 32, 0, nil, nil},
+	{"TABLE_COLLATION", mysql.TypeVarchar, 32, mysql.NotNullFlag, "utf8_general_ci", nil},
```
Comment (coocood) [inline infoschema/tables.go]: We don't support `utf8_general_ci`.

File: model/model_test.go [hunk 7]
```diff
@@ -72,7 +72,7 @@ func (*testModelSuite) TestModelBasic(c *C) {
 		ID:          1,
 		Name:        NewCIStr("t"),
 		Charset:     "utf8",
-		Collate:     "utf8",
+		Collate:     "utf8_general_ci",
```
Comment (coocood) [inline model/model_test.go]: utf8_bin

File: model/model_test.go [hunk 8]
```diff
@@ -83,7 +83,7 @@ func (*testModelSuite) TestModelBasic(c *C) {
 		ID:      1,
 		Name:    NewCIStr("test"),
 		Charset: "utf8",
-		Collate: "utf8",
+		Collate: "utf8_general_ci",
```
Comment (coocood) [inline model/model_test.go]: ditto

**Comment by shenli** [comment]
LGTM
@coocood PTAL

**Comment by zhexuany** [comment]
@coocood friendly ping.

File: infoschema/tables.go (hunk 2 above)
Comment (zimulala) [reply infoschema/tables.go]: Could we replace it with `mysql.DefaultCollationName`?

Comment (zhexuany) [reply infoschema/tables.go]: Yeah, but they are exactly same. - - 7

**Pull request merged**